test/controllers/pages_controller_test.rb

```diff
@@ -0,0 +1,7 @@
+require "test_helper"
+
+class PagesControllerTest < ActionDispatch::IntegrationTest
+  # test "the truth" do
+  #   assert true
+  # end
+end
```

Comment:
1. What do this file do?
2. What is `PagesControllerTest < ActionDispatch::IntegrationTest`?

Reply:
1. To run test for pages_controller. You'll add the tests further down the course
2. This, you would have to read up and let me know :D

Reply:
`ActionDispatch::IntegrationTest` is used specifically for integration testing where it groups a number of controllers needed and do testing for interfaces and flow of information between modules. Integration testing is needed to ensure the application work correctly when the code coded by different programmers are integrated.

`PagesControllerTest` inherits `ActionDispatch::IntegrationTest` so it can make use of some available helpers in `ActionDispatch::IntegrationTest` and no need to explicitly create the helpers itself.

---
Question:
1. Is integration testing needed for all parts of the software or only specific parts? If specific parts, which one?

Reply:
Software testing has 4 steps: unit testing, integration testing, system testing, and acceptance testing.

Unit testing test individual components, integration testing test component groups after unit testing is finished. Then system testing test the integrated system as a whole, then acceptance testing test the final system before it is release to the public.

After unit testing, integration testing comes in to test the individual components as a group to ensure the interfaces between components works correctly, for example, I have a `login` page and a `dashboard` page, we have tested that `login` page and `dashboard` page works fine with unit testing, now we proceed to testing the interface between `login` and `dashboard` page, we want the application to navigate to `dashboard` page after we successfully key in the correct information to login, so the testing of this process is the integration testing.

Reply:
> testing the interface between login and dashboard page

elaborate more on this

Reply:
We have 2 situations
1. `login` page is done but `dashboard` page hasn't
2. `dashboard` page is done but `login` page hasn't

`login` page is dependent on `dashboard` page. 
For the 1st situation, a dummy `dashboard` page is created, we call this stub(or called program), to test whether the application will proceed to the `dashboard` page when we input the correct information to login.
For the 2nd situation, a dummy `login` page is created, we call this driver (or calling program), we check whether the `dashboard` page can get to the correct `dashboard` page when we login with a dummy account, for example, I login using `dummy`, the `dashboard` page should display the `dashboard` page of `dummy` account.

Reply:
you are explaining as though (1) & (2) is different. But They are the same aren't they?

Page 1 that is done trying to test if can go to Page 2

???

Reply:
I'm also confused, for 1st one I think I get it, but for the driver I'm not so sure😅

Reply:
go read up more ==!

Reply:
> We have 2 situations
> 
> 1. `login` page is done but `dashboard` page hasn't
> 2. `dashboard` page is done but `login` page hasn't
> 
> `login` page is dependent on `dashboard` page. For the 1st situation, a dummy `dashboard` page is created, we call this stub(or called program), to test whether the application will proceed to the `dashboard` page when we input the correct information to login. For the 2nd situation, a dummy `login` page is created, we call this driver (or calling program), we check whether the `dashboard` page can get to the correct `dashboard` page when we login with a dummy account, for example, I login using `dummy`, the `dashboard` page should display the `dashboard` page of `dummy` account.

Correction: 
We total have 3 pages: `login`, `dashboard`, `logout` and only the `dashboard` page done, others are still in development. So we need a dummy `login` page to see whether the `dashboard` page is displaying the account user's dashboard page when we input the user ID from `login` page. So the `login` page is the `driver`. Now we have the `dashboard` page ready, we want to proceed to logout, let say by logging out, the `logout` page should display the user name who is logging out the page, so we need to pass the user id from the `dashboard` page to the `logout` page, but `logout` page is still under development, so we create a dummy `logout` page which we called a stub. So by creating drivers and stubs, we can do integrating testing without the need of the `login` and `logout` to be fully developed then only do the testing.

Take `dashboard` page as the one undergo integration testing, we need `login` page, that's why `login` page is a driver in the bottom-up approach, and `logout` page is a stub in the top-down approach.

---
Question
1. If according to the example above, does `login` page requires a driver and does `logout` page requires a stub?